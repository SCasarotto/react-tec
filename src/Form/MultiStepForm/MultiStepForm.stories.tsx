import React from 'react';

import { Story, Meta } from '@storybook/react';

import { Input } from '../Input';
import { MultiStepForm, MultiStepFormProps } from './MultiStepForm';

export default {
  title: 'Form/MultiStepForm',
  component: MultiStepForm,
  argTypes: {},
} as Meta;

const Template: Story<MultiStepFormProps> = (args) => (
  <MultiStepForm {...args} />
);

export const Default = Template.bind({});
Default.args = {
  onSubmit: () => {
    alert('Submitted!');
    return Promise.resolve();
  },
  stepData: [
    {
      title: 'Step 1',
      component: <Input onChange={(e) => console.log(e.target.value)} />,
      onNext: () => {
        console.log('Heading to step 2!');
        return Promise.resolve();
      },
    },
    {
      title: 'Step 22',
      component: <Input onChange={(e) => console.log(e.target.value)} />,
      onPrevious: () => {
        console.log('Heading back to step 1!');
        return Promise.resolve();
      },
      onNext: () => {
        console.log('Heading to step 3!');
        return Promise.resolve();
      },
    },
    {
      title: 'Step 333',
      component: <Input onChange={(e) => console.log(e.target.value)} />,
    },
  ],
};
