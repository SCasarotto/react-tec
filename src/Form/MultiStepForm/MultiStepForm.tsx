//
// NOTES:
//
// 10-11-18 SC
// - May be able to be improved by moving to a render prop form but seems fine for now.
// - Could also improve by adding react-transition-group around each step
//

import React, { useState } from 'react';

import { Form } from '../Form';
import {
  Container,
  StepContainer,
  StepBar,
  StepWrapper,
  StepNumber,
  StepTitle,
  ButtonContainer,
  StepButton,
} from './styledComponents';

export interface MultiStepFormStepData {
  title: string;
  component: React.ReactNode;
  // TODO: Try to make these types better
  onPrevious?(): Promise<any>;
  onNext?(): Promise<any>;
}
export interface MultiStepFormProps {
  className?: string;
  handleCancelOnClick?(): void;
  stepData: Array<MultiStepFormStepData>;
  // roundedButtons?: boolean
  // TODO: Try to make these types better
  onSubmit(): Promise<any>;
  onStepChange?(step: number): void;
  submitButtonId?: string;
}
/**
 *
 * Multiple step form container. `MultiStepForm` provides a quick and easy way to break a form into multiple parts and present the user with one part at a time.
 *
 */
export const MultiStepForm = React.forwardRef<
  HTMLFormElement,
  MultiStepFormProps
>((props, ref) => {
  const {
    className = '',
    handleCancelOnClick,
    stepData,
    // roundedButtons,
    submitButtonId,
    onStepChange,
    onSubmit,
  } = props;
  const stepCount = stepData?.length ?? 0;

  const [currentStep, setCurrentStep] = useState(0);

  const updateAndEmitStep = (step: number) => {
    setCurrentStep(step);
    if (onStepChange) {
      onStepChange(step);
    }
  };
  const handlePreviousPressed = async () => {
    const moveBack = () => {
      if (currentStep === 0) {
        if (handleCancelOnClick) {
          handleCancelOnClick();
        }
      } else {
        updateAndEmitStep(currentStep - 1);
      }
    };

    try {
      const onPrevious = stepData?.[currentStep]?.onPrevious;
      if (onPrevious) {
        await onPrevious();
        moveBack();
      } else {
        moveBack();
      }
    } catch (e) {
      console.log(e);
    }
  };
  const handleNextPressed = async () => {
    const moveForward = async () => {
      try {
        if (currentStep + 1 === stepData.length) {
          await onSubmit();
          updateAndEmitStep(0);
        } else {
          updateAndEmitStep(currentStep + 1);
        }
      } catch (e) {
        console.log(e);
      }
    };
    try {
      const onNext = stepData?.[currentStep]?.onNext;
      if (onNext) {
        await onNext();
        moveForward();
      } else {
        moveForward();
      }
    } catch (e) {
      console.log(e);
    }
  };

  return (
    <Container className={`tec-multiStepForm ${className}`}>
      <StepContainer className='tec-multiStepFormStepContainer'>
        <StepBar stepCount={stepCount} className='tec-multiStepFormStepBar' />
        {stepData?.map((step, index) => {
          const { title } = step;

          return (
            <StepWrapper
              key={step.title}
              stepCount={stepCount}
              className='tec-multiStepFormStepWrapper'
            >
              <StepNumber
                active={currentStep === index}
                className='tec-multiStepFormStepNumber'
              >
                {index + 1}
              </StepNumber>
              <StepTitle className='tec-multiStepFormStepTitle'>
                {title}
              </StepTitle>
            </StepWrapper>
          );
        })}
      </StepContainer>
      <Form className='tec-multiStepFormForm' ref={ref}>
        {stepData && stepData.length > 0 && stepData?.[currentStep]?.component}
        <ButtonContainer className='tec-multiStepFormButtonContainer'>
          {(handleCancelOnClick || currentStep > 0) && (
            <StepButton
              onClick={handlePreviousPressed}
              // position="left"
              // rounded={roundedButtons}
              className='tec-multiStepFormStepButton tec-multiStepFormStepButtonLeft'
            >
              {currentStep === 0 ? 'cancel' : 'previous'}
            </StepButton>
          )}
          <StepButton
            onClick={handleNextPressed}
            // position="right"
            // rounded={roundedButtons}
            // singleButton={!handleCancelOnClick && currentStep === 0}
            className='tec-multiStepFormStepButton tec-multiStepFormStepButtonright'
            id={submitButtonId}
          >
            {currentStep + 1 === stepCount ? 'submit' : 'next'}
          </StepButton>
        </ButtonContainer>
      </Form>
    </Container>
  );
});
