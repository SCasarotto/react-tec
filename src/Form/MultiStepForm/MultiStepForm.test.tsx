import React from 'react';

import { render, waitFor } from '@testing-library/react';

import { MultiStepForm } from './MultiStepForm';

describe('MultiStepForm', () => {
  // Renders Without Error
  it('matches snapshot', () => {
    const { asFragment } = render(
      <MultiStepForm stepData={[]} onSubmit={async () => jest.fn()} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  // Class Name
  it('accepts custom className', () => {
    const { container } = render(
      <MultiStepForm
        stepData={[]}
        onSubmit={async () => jest.fn()}
        className='myClassName'
      />,
    );
    expect(container.firstChild).toHaveClass('myClassName');
  });

  // Ref
  it('renders with ref', async () => {
    const ref: { current: HTMLFormElement | null } = { current: null };
    render(
      <MultiStepForm
        stepData={[]}
        onSubmit={async () => jest.fn()}
        ref={ref}
        data-testid='myMultiForm'
      />,
    );
    // Using this as the focus check didn't seem to work
    waitFor(() => expect(ref.current).toBeTruthy());
  });
});
