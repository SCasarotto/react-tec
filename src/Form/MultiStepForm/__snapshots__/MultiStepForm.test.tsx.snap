// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiStepForm matches snapshot 1`] = `
<DocumentFragment>
  .c2 {
  position: relative;
  width: 100%;
  margin: 0px auto;
  box-sizing: border-box;
}

.c4 {
  position: relative;
  display: block;
  width: 100%;
  font-size: 18px;
  padding: 5px 20px;
  text-align: center;
  cursor: pointer;
  border-radius: 5px;
  box-sizing: border-box;
  margin: 0px;
  border: 1px solid;
  -webkit-transition: color 0.2s ease-in,background-color 0.2s ease-in, border 0.2s ease-in;
  transition: color 0.2s ease-in,background-color 0.2s ease-in, border 0.2s ease-in;
}

.c4:hover,
.c4:active {
  border: 1px solid;
}

.c4:disabled {
  cursor: not-allowed;
  border: 1px solid;
}

.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  margin-bottom: 30px;
}

.c1 {
  position: absolute;
  top: 14px;
  height: 2px;
  width: -Infinity%;
  left: Infinity%;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

@media (max-width:500px) {
  .c0 {
    margin-bottom: 15;
  }
}

<div
    class="tec-multiStepForm "
  >
    <div
      class="c0 tec-multiStepFormStepContainer"
    >
      <div
        class="c1 tec-multiStepFormStepBar"
      />
    </div>
    <form
      class="c2 tec-form tec-multiStepFormForm"
    >
      <div
        class="c3 tec-multiStepFormButtonContainer"
      >
        <button
          class="c4 tec-button tec-multiStepFormStepButton tec-multiStepFormStepButtonright"
          type="button"
        >
          next
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
